### PR TITLE
Hold invites

### LINKED ISSUE #459: Send a "HOLD:" meeting invite to scheduling party until the meeting is confirmed.
Had someone schedule time on my calendar recently through Appointment and they noted that it would have been nice to have an invite sent to them to hold that time open until I confirmed the meeting. I agree. You could be waiting for confirmation and fill that slot without thinking and then end up with a conflict.

Some examples: We could simply make the title for the event "HOLD: Appointment - Ryan Sipes and (other party)" or an alternative could be "AWAITING CONFIRMATION: Appointment - Ryan Sipes and (other party)"

The invite could be updated with a new title (removing "HOLD:") once the event is confirmed.

### PR DESCRIPTION
**Pull request by devmount**
## Description of the Change

This change implements the [previously discussed](https://github.com/thunderbird/appointment/issues/459#issuecomment-2492915805) hold event flow.

This still needs heavy testing for both: Google and CalDAV.

## Applicable Issues

Fixes #459

### CODE REVIEW

File: backend/src/appointment/controller/apis/google_client.py [hunk 1]
```diff
@@ -141,19 +141,27 @@ def list_events(self, calendar_id, time_min, time_max, token):
 
         return items
 
-    def create_event(self, calendar_id, body, token):
+    def save_event(self, calendar_id, body, token):
```
Comment (devmount) [inline backend/src/appointment/controller/apis/google_client.py:144]: I renamed this cause we're now using it for creating and updating events.

File: backend/src/appointment/controller/apis/google_client.py [hunk 2]
```diff
@@ -141,19 +141,27 @@ def list_events(self, calendar_id, time_min, time_max, token):
...
                 raise EventNotCreatedException()
 
         return response
 
     def delete_event(self, calendar_id, event_id, token):
-        pass
+        response = None
+        with build('calendar', 'v3', credentials=token, cache_discovery=False) as service:
+            try:
+                response = service.events().delete(calendarId=calendar_id, eventId=event_id).execute()
```
Comment (devmount) [inline backend/src/appointment/controller/apis/google_client.py:159]: For some reason, this throws:

> [google_client.delete_event] Request Error: 404/[{'domain': 'global', 'reason': 'notFound', 'message': 'Not Found'}

Still investigating.

File: backend/src/appointment/controller/calendar.py [hunk 3]
```diff
@@ -501,7 +516,7 @@ def create_vevent(
             slot.start.replace(tzinfo=timezone.utc) + timedelta(minutes=slot.duration),
         )
         event.add('dtstamp', datetime.now(UTC))
-        event.add('status', 'CONFIRMED')
+        event.add('status', 'NEEDS-ACTION' if on_hold else 'CONFIRMED')
```
Comment (devmount) [inline backend/src/appointment/controller/calendar.py:519]: This needs to be tested directly in Thunderbird.

File: backend/src/appointment/controller/calendar.py [hunk 4]
```diff
@@ -840,3 +882,9 @@ def fix_caldav_urls(url: str) -> str:
             url = 'https://api.googlecontent.com/caldav/v2/'
 
         return url
+
+    @staticmethod
+    def default_event_title(slot: schemas.Slot, owner: schemas.Subscriber, prefix='') -> str:
```
Comment (devmount) [inline backend/src/appointment/controller/calendar.py:887]: TODO: add some comments

File: backend/src/appointment/database/repo/slot.py [hunk 5]
```diff
@@ -114,11 +114,3 @@ def delete(db: Session, slot_id: int):
...
     db.commit()
     return db_slot
-
-
-def is_available(db: Session, slot_id: int):
-    """check if slot is still available for booking"""
-    slot = get(db, slot_id)
-    if slot.schedule:
-        return slot and slot.booking_status == models.BookingStatus.requested
-    return False
```
Comment (devmount) [inline backend/src/appointment/database/repo/slot.py:124]: This isn't needed anymore since slots are now always unavailable as soon as they're booked.

**Review by devmount — COMMENTED** [review]
@MelissaAutumn This is still wip, but feel free to already take a look at it. Except for the remote event deletion, everythings already working fine on Google and CalDAV on my end.

File: backend/src/appointment/controller/apis/google_client.py (hunk 2 above)
Comment (MelissaAutumn) [reply backend/src/appointment/controller/apis/google_client.py:159]: That's very odd. We have the permissions to do so, so it shouldn't be that.

File: backend/src/appointment/controller/calendar.py (hunk 3 above)
Comment (devmount) [reply backend/src/appointment/controller/calendar.py:519]: Whoops, this was supposed to be `TENTATIVE`, but we can keep `CONFIRMED` as well. I don't think this is needed for the prompt to update. According to RFC 5546 and [Philipp](https://github.com/thunderbird/appointment/issues/459#issuecomment-2491213535), there needs to be a METHOD:REQUEST. I'll add that too.

File: backend/src/appointment/controller/apis/google_client.py (hunk 2 above)
Comment (MelissaAutumn) [reply backend/src/appointment/controller/apis/google_client.py:159]: Will do!

**Comment by MelissaAutumn** [comment]
Fixed the issue. It turns out Google needs their internal id field to delete events, and not the one we create via their import method. So once we save an event it returns the event object back to us, so I just store it on appointment.external_id (new field!). CalDAV doesn't need to use this field, so it falls back to uuid. 

Also discovered that it may fail if the user deletes their event before they reject the request. I'm not sure if we want to handle that any differently thank a "link is invalid" error. Additionally I discovered if you switch the schedule's calendar, delete and update requests were being attempted on the schedule's calendar and not the appointment's calendar (the one it's actually on.) so I patched that up. 

Feel free to give it a try again to see if it all works on your end now!

File: backend/src/appointment/controller/apis/google_client.py (hunk 2 above)
Comment (devmount) [reply backend/src/appointment/controller/apis/google_client.py:161]: Tested, works now 🎉

**Review by devmount — COMMENTED** [review]
Excellent work @MelissaAutumn 👏🏻  Google event deletion on denial works flawlessly now on my end too!

> Also discovered that it may fail if the user deletes their event before they reject the request. I'm not sure if we want to handle that any differently thank a "link is invalid" error.

I think it's fine to keep that error for now and see, if we want to maybe make that more specific in the future if we need it.

> Additionally I discovered if you switch the schedule's calendar, delete and update requests were being attempted on the schedule's calendar and not the appointment's calendar (the one it's actually on.) so I patched that up.

Oh, thank you! Didn't think of that particular case. Maybe we should also inform the user about open bookings when switching calendars on the schedule page or disabling them in the settings? Just as idea to improve the UX at that point, but out of scope here.

Imo we can merge this now.

**Review by MelissaAutumn — APPROVED** [review]

**Pull request merged**